bokeh/charts/builder.py

```diff
@@ -547,6 +552,33 @@ def generate_help(cls):
 
         return help_str
 
+    @staticmethod
+    def _sort_legend(sort_legend, legends, attributes):
+        """Sort legends sorted by looping though sort_legend items (
+        see :attr:`Builder.sort_legend` for more details)
+        """
+        if len(sort_legend) > 0:
+            for attr, asc in sort_legend:
+                if len(attributes[attr].columns) > 0:
+
+                    # TODO(fpliger): attributes should be consistent and not
+                    #               need any type checking but for
+                    #               the moment it is not, specially when going
+                    #               though a process like binning or when data
+                    #               is built for HeatMap, Scatter, etc...
+                    item_order = [x[0] if isinstance(x, tuple) else x
+                        for x in attributes[attr].items]
+
+                    item_order = [str(x) if not isinstance(x, string_types)
+                        else x for x in item_order]
+
+                    def foo(leg):
+                        return item_order.index(leg[0])
+
+                    return list(sorted(legends, key=foo, reverse=(not asc)))
```

Comment:
Right I'm not suggesting implementing sub-sorting is the thing to do, only that the current interface suggests something more general than is actually implemented, so maybe the property needs to be tightened up. `Tuple(String, Bool)`  (or maybe better with enum: `Tuple(String, SortDirection)`) seems reasonable to me.